### PR TITLE
Drop node < 4 support

### PR DESCRIPTION
**Pull request by shvaikalesh**

### CODE REVIEW

File: test/spec.js [hunk 1]
```diff
@@ -751,7 +754,7 @@ describe('checkDependencies', () => {
             consoleLogStub.restore();
             consoleErrorStub.restore();
 
-            Reflect.deleteProperty(process, 'exitCode');
+            delete process.exitCode;
```
Comment (shvaikalesh) [inline test/spec.js]: Whatever eslint says, `delete` is by far not equal to `Reflect.deleteProperty`, it is equavialent to
```js
if (Reflect.deleteProperty(obj, 'key')) throw new TypeError
```

**Comment by paulmillr** [comment]
@shvaikalesh could you fix tests? We'll use the fork for now.

**Comment by paulmillr** [comment]
@mgol I don't see how this is useful either — 0.12 is no longer supported; and supporting 0.12 adds additional 5mb to [Brunch](http://brunch.io) installation which is increasing the size by 40%. Because, Babel and stuff.

**Comment by shvaikalesh** [comment]
According to `README.md` and `engines` in `package.json`, this projects does not support Node.js `0.12`. This PR removes extra dependencies.

File: appveyor.yml [hunk 2]
```diff
@@ -21,6 +21,7 @@ install:
   # Log HTTP requests.
   - npm config set loglevel http
   # Typical npm stuff.
+  - npm install -g npm@3
```
Comment (mgol) [inline appveyor.yml]: I'd rather not add this line. Node 6 comes with npm 3 anyway and I'd like this project to keep working in the defautl Node 4 configuration, i.e. with the built-in npm 2.

File: bin/cli.js [hunk 3]
```diff
@@ -3,17 +3,17 @@
...
 const minimist = require('minimist');
-const _ = require('lodash');
+const camelCase = require('lodash.camelcase');
 const checkDependencies = require('../lib/check-dependencies');
 
 const argv = minimist(process.argv.slice(2));
 
 // camelCase the options
-for (const key in argv) {
+Object.keys(argv).forEach(key => {
```
Comment (mgol) [inline bin/cli.js]: Why this change? This should work in Node 4 as it was.

File: appveyor.yml [hunk 4]
```diff
@@ -37,6 +37,7 @@ build: off
 matrix:
 #  fast_finish: true
   allow_failures:
+    - nodejs_version: "4"
```
Comment (mgol) [inline appveyor.yml]: I see that Node 6 tests seem to pass on AppVeyor, it's just Node 4 that's problematic. Maybe we could disable allowing failures in Node 6?

**Comment by mgol** [comment]
I was thinking about de-facto supporting Node 0.12 for a few next months (its upstream support ends in December) while not "advertising" that support in README and the `package.json` `engines` field. Was that more confusing than I thought?

I'm not opposed to formally dropping the support now since I see it matters to you. :)

File: bin/cli.js (hunk 3 above)
Comment (shvaikalesh) [inline bin/cli.js]: `for in` is dangerous: it is slow and behaves differently in older V8 and newest V8 (will be shipped in Node 7)/other browsers.

**Comment by mgol** [comment]
I wonder if it would be possible to:

1. Leave code fully using ES6, just without the `Reflect` stuff.
2. Transpile only features not available in Node 4 (like spread); e.g. via the [babel-preset-es2015-node4](https://www.npmjs.com/package/babel-preset-es2015-node4) plugin.
3. Since the features missing in Node 4 are mostly syntactic, transpile it via Babel but without adding `babel-runtime` to dependencies, hence not increasing the final package size.

What do you think?

File: bin/cli.js (hunk 3 above)
Comment (mgol) [inline bin/cli.js]: Performance shouldn't be a problem for `argv`, this list won't be very large.

What are the differences in behavior? Is it about the iteration order or something else?

Comment (shvaikalesh) [inline bin/cli.js]: It is about enumerable properties being shadowed by non-enumerables and skipped. That would be a concern, mainly, if we would need prototype chain, but we don't. That's why I purpose this change: using `for in` is kinda overkill here.

**Comment by shvaikalesh** [comment]
> Transpile only features not available in Node 4

That is nice, but I personally, would not bother:

`...` operator can improve one test, removing `.apply`
destructuring can improve `checkPackage` helper a bit

File: bin/cli.js (hunk 3 above)
Comment (mgol) [inline bin/cli.js]: I guess my main point is that for simple objects where neither performance nor complex prototype chain handling matters `for (const key in argv) {` reads nicer than the `Object.keys(argv).forEach(key => {` counterpart.

There's a third option as well:
```js
for (const key of Object.keys(argv)) {
    /* ... */
}
```

If we're talking performance, it should be even faster. ;)

**Comment by mgol** [comment]
> `...` operator can improve one test, removing .apply
> destructuring can improve `checkPackage` helper a bit

OK, I see how that may be a little overkill here then.

File: bin/cli.js (hunk 3 above)
Comment (shvaikalesh) [inline bin/cli.js]: I agree on "reads better". On performance: theoretically, it should be faster (because no `[[HasProperty]]` checks and function calls), but it is not yet optimized by V8 as good as good old `forEach`. Changed to `for of`.

**Comment by mgol** [comment]
Since Babel is out, please make sure there are no mentions of removed files/directories left. For example, I see `.gitignore` still mentions `dist` and `package.json`'s `files` field contains `dist`, `index.js` etc.

**Comment by mgol** [comment]
@shvaikalesh I'll need to bump a minor or major after landing it since it disables Node 0.12 support for real this time so it's a breaking change. Do you think this project is stable/advanced enough to warrant bumping to `1.0.0`?

**Comment by shvaikalesh** [comment]
@mgol After brief examination of commit history, API changes & issues, I would say `1.0` is the way to go.

**Comment by mgol** [comment]
Thanks for the PR! `1.0.0` published to npm.

**Comment by mgol** [comment]
For the record, I've been struggling with AppVeyor failures for some time now; it used to work with Node 0.12 but not io.js and newer Nodes. I reported it at http://help.appveyor.com/discussions/problems/2532-invalid-npm-config-on-iojs-2-images but I haven't had time to investigate it properly to pinpoint the issue on a smaller project. I'm happy this problem is gone in Node 6 (as it seems) but Node 4 tests on Windows remain broken.
